File: aftermath-sdk-ruby/aftermath_middleware.rb
require 'json'
require 'net/http'
require 'uri'
require 'securerandom'

module Aftermath
  class Middleware
    def initialize(app, options = {})
      @app = app
      @collector_url = options[:collector_url] || 'http://localhost:8090/api/v1/incidents'
      @service_name = options[:service_name] || 'ruby-service'
    end

    def call(env)
      begin
        @app.call(env)
      rescue => exception
        capture_incident_async(env, exception)
        raise exception # Re-throw after capturing
      end
    end

    private

    def capture_incident_async(env, exception)
      Thread.new do
        begin
          headers = {}
          sensitive = ['authorization', 'cookie', 'x-api-key']
          env.each do |k, v|
            if k.start_with?('HTTP_')
              header_name = k[5..-1].split('_').map(&:capitalize).join('-')
              if sensitive.include?(header_name.downcase)
                headers[header_name] = '[REDACTED]'
              else
                headers[header_name] = v.to_s
              end
            end
          end

          event = {
            incidentId: SecureRandom.uuid,
            traceId: env['HTTP_X_TRACE_ID'] || SecureRandom.uuid,
            timestamp: (Time.now.to_f * 1000).to_i,
            request: {
              method: env['REQUEST_METHOD'] || 'GET',
              uri: env['REQUEST_URI'] || '/',
              headers: headers,
              body: '',
              timestamp: (Time.now.to_f * 1000).to_i
            },
            error: {
              exceptionClass: exception.class.name,
              message: exception.message,
              stackTrace: exception.backtrace ? exception.backtrace.join("\n") : '',
              statusCode: 500
            },
            deployment: {
              serviceName: @service_name,
              serviceVersion: '1.0.0',
              environment: 'production',
              commitHash: 'ruby-commit'
            }
          }

          uri = URI.parse(@collector_url)
          http = Net::HTTP.new(uri.host, uri.port)
          http.open_timeout = 2
          http.read_timeout = 2
          req = Net::HTTP::Post.new(uri.request_uri, { 'Content-Type' => 'application/json' })
          req.body = event.to_json
          http.request(req)
        rescue => e
          # Fail-open protection
        end
      end
    end
  end
end

File: aftermath-sdk-php/AftermathMiddleware.php
<?php

namespace Aftermath\Sdk;

class AftermathMiddleware {
    private string $collectorUrl;
    private string $serviceName;

    public function __construct(string $collectorUrl = "http://localhost:8090/api/v1/incidents", string $serviceName = "php-service") {
        $this->collectorUrl = $collectorUrl;
        $this->serviceName = $serviceName;
    }

    public function handle(\Throwable $exception, array $serverParams = []): void {
        try {
            $sensitive = ["authorization", "cookie", "x-api-key", "http_authorization", "http_cookie"];
            $headers = [];
            foreach ($serverParams as $k => $v) {
                if (str_starts_with($k, 'HTTP_')) {
                    $headerName = str_replace(' ', '-', ucwords(strtolower(str_replace('_', ' ', substr($k, 5)))));
                    if (in_array(strtolower($k), $sensitive)) {
                        $headers[$headerName] = "[REDACTED]";
                    } else {
                        $headers[$headerName] = (string)$v;
                    }
                }
            }

            $event = [
                "incidentId" => sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x', mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0x0fff) | 0x4000, mt_rand(0, 0x3fff) | 0x8000, mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)),
                "traceId" => $serverParams['HTTP_X_TRACE_ID'] ?? sprintf('%04x%04x-%04x-%04x-%04x-%04x%04x%04x', mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0x0fff) | 0x4000, mt_rand(0, 0x3fff) | 0x8000, mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)),
                "timestamp" => (int)(microtime(true) * 1000),
                "request" => [
                    "method" => $serverParams['REQUEST_METHOD'] ?? 'GET',
                    "uri" => $serverParams['REQUEST_URI'] ?? '/',
                    "queryParams" => [],
                    "headers" => $headers,
                    "body" => "",
                    "timestamp" => (int)(microtime(true) * 1000)
                ],
                "error" => [
                    "exceptionClass" => get_class($exception),
                    "message" => $exception->getMessage(),
                    "stackTrace" => $exception->getTraceAsString(),
                    "statusCode" => 500
                ],
                "deployment" => [
                    "serviceName" => $this->serviceName,
                    "serviceVersion" => "1.0.0",
                    "environment" => "production",
                    "commitHash" => "php-commit"
                ]
            ];

            $this->sendAsync(json_encode($event));
        } catch (\Throwable $e) {
            // Fail-open protection
        }
    }

    private function sendAsync(string $json): void {
        $ch = curl_init($this->collectorUrl);
        curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
        curl_setopt($ch, CURLOPT_POSTFIELDS, $json);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 2);
        curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/json']);
        curl_exec($ch);
        curl_close($ch);
    }
}
